src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -0,0 +1,397 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, VecDeque};
+
+use bytes::Bytes;
+use risingwave_common::catalog::TableId;
+use risingwave_common::hash::VirtualNode;
+use risingwave_hummock_sdk::key::FullKey;
+use risingwave_hummock_sdk::table_watermark::{ReadTableWatermark, WatermarkDirection};
+
+use crate::hummock::iterator::{Forward, HummockIterator};
+use crate::hummock::value::HummockValue;
+use crate::hummock::HummockResult;
+use crate::monitor::StoreLocalStatistic;
+
+pub struct SkipWatermarkIterator<I> {
+    inner: I,
+    watermarks: BTreeMap<TableId, ReadTableWatermark>,
+    remain_watermarks: VecDeque<(TableId, VirtualNode, WatermarkDirection, Bytes)>,
```

Comment:
> https://github.com/risingwavelabs/risingwave/pull/13429#discussion_r1405825993
https://github.com/risingwavelabs/risingwave/pull/13429#discussion_r1405827053

It's for compaction because in compaction we have handle multiple tables together.